extra/DEVELOPER_DOCS/Code Conventions.md

```diff
@@ -0,0 +1,546 @@
+| `Vocab`    | `vocab`               | `vocab = Vocab()`                           |
+| `Example`  | `example`, `eg`       | `example = Example.from_dict(doc, gold)`    |
+| `Config`   | `config`, `cfg`       | `config = Config().from_disk("config.cfg")` |
+
+We try to avoid introducing too many temporary variables, as these clutter your namespace. It's okay to re-assign to an existing variable, but only if the value has the same type.
+
+```diff
+ents = get_a_list_of_entities()
+ents = [ent for ent in doc.ents if ent.label_ == "PERSON"]
+- ents = {(ent.start, ent.end): ent.label_ for ent in ents}
++ ent_mappings = {(ent.start, ent.end): ent.label_ for ent in ents}
+```
+
+### Naming methods and functions
+
+Try choosing short and descriptive names wherever possible and imperative verbs for methods that do something, e.g. `disable_pipes`, `add_patterns` or `get_vector`. Private methods and functions that are not intended to be part of the user-facing API should be prefixed with an underscore `_`. It's often helpful to look at the existing classes for inspiration.
+
+Objects that can be serialized, e.g. data structures and pipeline components, should implement the same consistent methods for serialization. Those usually include at least `to_disk`, `from_disk`, `to_bytes` and `from_bytes`. Some objects can also implement more specific methods like `{to/from}_dict` or `{to/from}_str`.
+
+## Error handling
+
+We always encourage writing helpful and detailed custom error messages for everything we can anticipate going wrong, and including as much detail as possible. spaCy provides a directory of error messages in `errors.py` with unique codes for each message. This allows us to keep the code base more concise and avoids long and nested blocks of texts throughout the code that disrupt the reading flow. The codes make it easy to find references to the same error in different places, and also helps identify problems reported by users (since we can just search for the error code).
+
+Errors can be referenced via their code, e.g. `Errors.E123`. Messages can also include placeholders for values, that can be populated by formatting the string with `.format()`.
+
+```python
+class Errors:
+    E123 = "Something went wrong"
+    E456 = "Unexpected value: {value}"
+```
+
+```diff
+if something_went_wrong:
+-    raise ValueError("Something went wrong!")
++    raise ValueError(Errors.E123)
+
+if not isinstance(value, int):
+-    raise ValueError(f"Unexpected value: {value}")
++    raise ValueError(Errors.E456.format(value=value))
+```
+
+As a general rule of thumb, all error messages raised within the **core library** should be added to `Errors`. The only place where we write errors and messages as strings is `spacy.cli`, since these functions typically pretty-print and generate a lot of output that'd otherwise be very difficult to separate from the actual logic.
+
+### Re-raising exceptions
```

Comment:
Nervous about mentioning this, as it can cause a lot of trouble. I think we use this much too often ourselves, so I'm reluctant to encourage people to author code with it.

Reply:
This section is about re-raising exceptions in general, though, and I think the way we're using it is fine? Re-raising from an existing exception is a really good pattern IMO and has zero impact on the traceback. It only changes the wording of how the re-raised exception is presented in Python ("...was the direct cause..."). I think this is very good and useful.

Re-raising from `None` is more problematic if it's used incorrectly, but I don't think we're really doing that. We previously had a problem where we were polluting people's traceback with chained errors that were not relevant to the exception we were re-raising. Maybe we just need a more explicit warning here that it should only be used for exceptions we can very clearly anticipate, and that it requires careful error messages.